optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,353 @@
+            first_trial = min(past_trials, key=lambda t: t._trial_id)
+            self._initial_search_space = self._infer_initial_search_space(first_trial)
+            return self._initial_search_space
+
+    def _infer_initial_search_space(self, trial: FrozenTrial) -> Dict[str, BaseDistribution]:
+
+        search_space = OrderedDict()  # type: OrderedDict[str, BaseDistribution]
+        for param_name, distribution in trial.distributions.items():
+            if not isinstance(distribution, _NUMERICAL_DISTRIBUTIONS):
+                continue
+            search_space[param_name] = distribution
+
+        return search_space
+
+    def _log_independent_sampling(self, trial: FrozenTrial, param_name: str) -> None:
+        _logger.warning(
+            "The parameter '{}' in trial#{} is sampled independently "
+            "by using `{}` instead of `QMCSampler` "
+            "(optimization performance may be degraded). "
+            "`QMCSampler` does not support dynamic search space or `CategoricalDistribution`. "
+            "You can suppress this warning by setting `warn_independent_sampling` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this independent sampling is intended behavior.".format(
+                param_name, trial.number, self._independent_sampler.__class__.__name__
+            )
+        )
+
+    @staticmethod
+    def _log_asyncronous_seeding() -> None:
+        _logger.warning(
+            "No seed is provided for `QMCSampler` and the seed is set randomly. "
+            "If you are running multiple `QMCSampler`s in parallel and/or distributed "
+            " environment, the same seed must be used in all samplers to ensure that resulting "
+            "samples are taken from the same QMC sequence. "
+            "You can suppress this warning by setting `warn_asyncronous_seeding` "
+            "to `False` in the constructor of `QMCSampler`, "
+            "if this random seeding is intended behavior."
+        )
+
+    def sample_independent(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        param_name: str,
+        param_distribution: BaseDistribution,
+    ) -> Any:
+
+        if self._initial_search_space is not None:
+            if self._warn_independent_sampling:
+                self._log_independent_sampling(trial, param_name)
+
+        return self._independent_sampler.sample_independent(
+            study, trial, param_name, param_distribution
+        )
+
+    def sample_relative(
+        self, study: Study, trial: FrozenTrial, search_space: Dict[str, BaseDistribution]
+    ) -> Dict[str, Any]:
+
+        if search_space == {}:
+            return {}
+
+        sample = self._sample_qmc(study, trial, search_space)
+        trans = _SearchSpaceTransform(search_space)
+        sample = scipy.stats.qmc.scale(sample, trans.bounds[:, 0], trans.bounds[:, 1])
+        sample = trans.untransform(sample[0, :])
+
+        return sample
+
```

Comment:
I found something like this in other samplers but not sure if it is necessary here... 
```suggestion

    def after_trial(
        self,
        study: "optuna.Study",
        trial: "optuna.trial.FrozenTrial",
        state: TrialState,
        values: Optional[Sequence[float]],
    ) -> None:

        self._independent_sampler.after_trial(study, trial, state, values)
        
```

Reply:
We'll probably add it for consistency sooner or later but you can leave it out if you want.